Source/Core/VideoBackends/D3D12/D3DCommandListManager.cpp

```diff
@@ -0,0 +1,352 @@
+
+	m_dirty_pso = true;
+	m_dirty_vertex_buffer = true;
+	m_dirty_ps_cbv = true;
+	m_dirty_vs_cbv = true;
+	m_dirty_gs_cbv = true;
+	m_dirty_samplers = true;
+	m_current_topology = D3D_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP;
+}
+
+void D3DCommandListManager::GetCommandList(ID3D12GraphicsCommandList** command_list) const
+{
+#ifdef USE_D3D12_QUEUED_COMMAND_LISTS
+	*command_list = this->m_queued_command_list;
+#else
+	*command_list = this->m_backing_command_list;
+#endif
+}
+
+void D3DCommandListManager::ProcessQueuedWork()
+{
+	m_queued_command_list->ProcessQueuedItems();
+}
+
+void D3DCommandListManager::ExecuteQueuedWork(bool wait_for_gpu_completion)
+{
+	if (wait_for_gpu_completion)
+	{
+		m_queue_fence_value++;
+	}
+
+#ifdef USE_D3D12_QUEUED_COMMAND_LISTS
+	CheckHR(m_queued_command_list->Close());
+	m_queued_command_list->QueueExecute();
+
+	if (wait_for_gpu_completion)
+	{
+		m_queued_command_list->QueueFenceGpuSignal(m_queue_fence, m_queue_fence_value);
+	}
+
+	ResetCommandListWithIdleCommandAllocator();
+
+	m_queued_command_list->ProcessQueuedItems();
+#else
+	CheckHR(m_backing_command_list->Close());
+
+	ID3D12CommandList* const commandListsToExecute[1] = { m_backing_command_list };
+	m_command_queue->ExecuteCommandLists(1, commandListsToExecute);
+
+	if (wait_for_gpu_completion)
+	{
+		CheckHR(m_command_queue->Signal(m_queue_fence, m_queue_fence_value));
+	}
+
+	if (m_current_command_allocator == 0)
+	{
+		PerformGpuRolloverChecks();
+	}
+
+	ResetCommandListWithIdleCommandAllocator();
+#endif
+
+	SetInitialCommandListState();
+
+	if (wait_for_gpu_completion)
+	{
+		WaitOnCPUForFence(m_queue_fence, m_queue_fence_value);
+	}
+}
+
+void D3DCommandListManager::ExecuteQueuedWorkAndPresent(IDXGISwapChain* swap_chain, UINT sync_interval, UINT flags)
+{
+#ifdef USE_D3D12_QUEUED_COMMAND_LISTS
+	CheckHR(m_queued_command_list->Close());
+	m_queued_command_list->QueueExecute();
+	m_queued_command_list->QueuePresent(swap_chain, sync_interval, flags);
+	m_queued_command_list->ProcessQueuedItems();
+
+	if (m_current_command_allocator == 0)
+	{
+		PerformGpuRolloverChecks();
+	}
+
+	m_current_command_allocator = (m_current_command_allocator + 1) % m_command_allocator_lists[m_current_command_allocator_list].size();
+
+	ResetCommandListWithIdleCommandAllocator();
+
+	SetInitialCommandListState();
+#else
+	ExecuteQueuedWork();
+	CheckHR(swap_chain->Present(sync_interval, flags));
+#endif
+}
+
+void D3DCommandListManager::WaitForQueuedWorkToBeExecutedOnGPU()
+{
+	// Wait for GPU to finish all outstanding work.
+	m_queue_fence_value++;
+#ifdef USE_D3D12_QUEUED_COMMAND_LISTS
+	m_queued_command_list->QueueExecute();
+	m_queued_command_list->QueueFenceGpuSignal(m_queue_fence, m_queue_fence_value);
+
+	m_queued_command_list->ProcessQueuedItems();
+#else
+	CheckHR(m_command_queue->Signal(m_queue_fence, m_queue_fence_value));
+#endif
+	WaitOnCPUForFence(m_queue_fence, m_queue_fence_value);
+}
+
+void D3DCommandListManager::PerformGpuRolloverChecks()
+{
```

Comment:
Correct, the frame latency is two, which is enough for maximum CPU/GPU pipelining. Can render at upwards of 1000fps..